framework/src/systems/NonlinearSystemBase.C

```diff
@@ -2587,8 +2587,12 @@ NonlinearSystemBase::setVariableGlobalDoFs(const std::string & var_name)
   AllLocalDofIndicesThread aldit(_sys.system(), {var_name});
   ConstElemRange & elem_range = *_mesh.getActiveLocalElementRange();
   Threads::parallel_reduce(elem_range, aldit);
-  _communicator.set_union(aldit._all_dof_indices);
-  _var_all_dof_indices.assign(aldit._all_dof_indices.begin(), aldit._all_dof_indices.end());
+
+  // Gather the dof indices across procs to get all the dof indices for var_name
+  aldit.dofIndicesSetUnion();
```

Comment:
I am concerned  on this

Reply:
The name of this method is `setVariableGlobalDoFs`. Sounds like a global operation to me. Look at the deleted lines. We were already doing a set union before. Now it's just encapsulated. This is the only place we do global communication. All the other `getDofIndices` calls are for local dofs only, e.g. we do not perform a global set union like we do here.

Reply:
OK, the code is called from here:

```
  if (_requires_nonlocal_coupling)
    setVariableAllDoFMap(_uo_jacobian_moose_vars[0]);
```

It may be fine as long as `_requires_nonlocal_coupling=false`.

Reply:
It is false almost all the time. I think virtually none of our users are using the non-local capability.

But this does beg the question of whether we could build up the global dof indices for a variable without doing communication.